### DISCUSSION TITLE
Keycloak add custom cookie to redirect URL

### DISCUSSION THREAD

**Discussion by mshanah** [keycloak/keycloak#13240] [Q&A]
we have legacy Application that authenticate by validating specific cookie to table storing all the cookies with there expired time , how can i achieve same using keycloak? can i write custom code so after the user successfully logged in i set this cookie when the keycloak send the redirect URL?

**Comment by jossemarGT** [comment]
Hey @mshanah by any chance did you find to accomplish this? I'm about to dive into the code/docs but it is cheaper to ask :bad poker face:

**Reply by mshanah** [reply]
@jossemarGT am sorry i cant find a way to do it till now, please if you find any way to help please

**Reply by jossemarGT** [reply]
No worries @mshanah . I found the "Authentication SPI" section[0] in the documentation and this custom authentication SPI example[1]. So, I _think_ those resources can come handy to implement what we both need. 

I cannot promise anything but if I manage to create something in my spare time, I'll surely let you know.

Happy hacking! :octocat: 

[0] https://www.keycloak.org/docs/latest/server_development/#_auth_spi
[1] https://github.com/keycloak/keycloak/tree/main/examples/providers/authenticator

**Reply by mshanah** [reply]
i can help i create some spi but i could not no which one to use , if you can check which one we should use that will execute after successfully authenticated .
thanks

**Reply by jossemarGT** [reply]
Hmm... This example class[0] _might_ be what you are looking for. However, I still don't have a good understanding on what other Classes must be implemented to fulfill  to get a minimal SPI, neither how to compile them. 

In other words. I still think the example+docs is a better guide than me, since there's plenty experimentation rounds on my end. :octocat: 

[0] https://github.com/keycloak/keycloak/blob/main/examples/providers/authenticator/src/main/java/org/keycloak/examples/authenticator/SecretQuestionAuthenticator.java

**Comment by vplemyannik** [comment]
Hi, i have faced with such problem.
i've been implementing the custom OIDC mapper to add custom claims into the JWT token, and attach legacy cookies to http response. 
```java
 @Override
    public AccessToken transformAccessToken(
            AccessToken token,
            ProtocolMapperModel mappingModel,
            KeycloakSession keycloakSession,
            UserSessionModel userSession,
            ClientSessionContext clientSessionCtx) {

        // 1. Do http request to your legacy application
        // 2. Extract  cookies from response
        // 3. Attach those cookies to keycloak response 
        HttpResponse originalResponse = ResteasyProviderFactory.getInstance().getContextData(HttpResponse.class);
        originalResponse.addNewCookie(new NewCookie(
                "YOUR COOKIE KEY",
                "YOUR COOKIE VALUE",
                "/",
                null,
                null,
                NewCookie.DEFAULT_MAX_AGE,
                false,
                true));
        setClaim(token, mappingModel, userSession, keycloakSession, clientSessionCtx);
        return token;
    }
```

**Reply by jossemarGT** [reply]
Hey! That's good to know, thanks for sharing @vplemyannik !

**Reply by vplemyannik** [reply]
Another workaround is - implement your own Authenticator. It more suitable place for these changes because you want only attach cookies, but OIDC mapper is more about token transformation rather than http modification.
```java
public class CustomAuthenticator extends UsernamePasswordForm {

    @Override
    public void authenticate(AuthenticationFlowContext context) {
        // 1. Do http request to your legacy application
        // 2. Extract  cookies from response
        // 3. Attach those cookies to keycloak response 
        HttpResponse originalResponse = ResteasyProviderFactory.getInstance().getContextData(HttpResponse.class);
        originalResponse.addNewCookie(new NewCookie(
                "YOUR COOKIE KEY",
                "YOUR COOKIE VALUE",
                "/",
                null,
                null,
                NewCookie.DEFAULT_MAX_AGE,
                false,
                true));
        super.authenticate(context);
    }
}
```

**Reply by mshanah** [reply]
i build Authenticator and added the custom cookie there, but  if i have flow where after authentication i have OTP  then the cookie will be set even when the OTP is not yet completed.
so i don't know where is the right place or SPI that will trigger after all authentication flow finish :(

**Reply by vplemyannik** [reply]
@mshanah  as i remember there is sorting for flow. I mean you can put your authenticator in the bottom of the flow list. In this way your authenticator will be executed only if all previous flows are finished successfully  
But you need to create a new auth flow from scratch 
![image](https://user-images.githubusercontent.com/17518385/199433572-8fc03c68-c9e4-4ce2-b260-5510ed48c4d1.png)


![image](https://user-images.githubusercontent.com/17518385/199432791-54443ce0-2cf6-412d-922f-eff97175919c.png)

If you don't want to configure auth flow i suggest to your implement cookies attachment through OIDC mapper, because it will be automatically invoked for all flows and only when all auth actions have passed successfully

**Reply by mshanah** [reply]
yes , thanks.
i will try it
